loopback/index.d.ts

```diff
@@ -0,0 +1,2585 @@
+
+      /**
+      * Compare if two principals are equal
+      * Returns true if argument principal is equal to this principal.
+      * @param {any} p The other principal
+      */
+
+      equals(p: any): void;
+
+      /**
+      * A request to access protected resources.
+      * @param {string} model The model name
+      * @param {string} property
+      * @param {string} accessType The access type
+      * @param {string} permission The requested permission
+      * @returns {AccessRequest}
+      * @class
+      */
+
+}
+
+      declare class AccessRequest {
+
+      /**
+      * Does the request contain any wildcards?
+      *
+      * @returns {boolean}
+      */
+
+      isWildcard(): boolean;
+
+      /**
+      * Does the given `ACL` apply to this `AccessRequest`.
+      *
+      * @param {ACL} acl
+      */
+
+      exactlyMatches(acl: ACL): void;
+
+      /**
+      * Is the request for access allowed?
+      *
+      * @returns {boolean}
+      */
+
+      isAllowed(): boolean;
+
+}
+
+/**
+* The base class for **all models**.
+*
+* **Inheriting from `Model`**
+*
+* ```js
+* var properties = {...};
+* var options = {...};
+* var MyModel = loopback.Model.extend('MyModel', properties, options);
+* ```
+*
+* **Options**
+*
+*  - `trackChanges` - If true, changes to the model will be tracked. **Required
+* for replication.**
+*
+* **Events**
+*
+* #### Event: `changed`
+*
+* Emitted after a model has been successfully created, saved, or updated.
+* Argument: `inst`, model instance, object
+*
+* ```js
+* MyModel.on('changed', function(inst) {
+*   console.log('model with id %s has been changed', inst.id);
+*   // => model with id 1 has been changed
+* });
+* ```
+*
+* #### Event: `deleted`
+*
+* Emitted after an individual model has been deleted.
+* Argument: `id`, model ID (number).
+*
+* ```js
+* MyModel.on('deleted', function(id) {
+*   console.log('model with id %s has been deleted', id);
+*   // => model with id 1 has been deleted
+* });
+* ```
+*
+* #### Event: `deletedAll`
+*
+* Emitted after an individual model has been deleted.
+* Argument: `where` (optional), where filter, JSON object.
+*
+* ```js
+* MyModel.on('deletedAll', function(where) {
+*   if (where) {
+*     console.log('all models where ', where, ' have been deleted');
+*     // => all models where
+*     // => {price: {gt: 100}}
+*     // => have been deleted
+*   }
+* });
+* ```
+*
+* #### Event: `attached`
+*
+* Emitted after a `Model` has been attached to an `app`.
+*
+* #### Event: `dataSourceAttached`
+*
+* Emitted after a `Model` has been attached to a `DataSource`.
+*
+* #### Event: set
+*
+* Emitted when model property is set.
+* Argument: `inst`, model instance, object
+*
+* ```js
+* MyModel.on('set', function(inst) {
+*   console.log('model with id %s has been changed', inst.id);
+*   // => model with id 1 has been changed
+* });
+* ```
+*
+* @param {any} data
+* @property {string} Model.modelName The name of the model. Static property.
```

Comment:
JSDoc mentions properties that aren't present in the class declaration.
Would be better to convert all this into TypeScript. e.g.:
```ts
declare class Model {
    /** The name of the model. */
    static modelName: string;
    ...
}
```